### DISCUSSION TITLE
Controllers remounted

### DISCUSSION THREAD

**Discussion by Asadbek13** [react-querybuilder/react-querybuilder#357] [Q&A] [answered]
I wanted to know in which case the Controllers of the query builder or even the query builder itself is remounted?

**Comment by jakeboone02** [comment, marked as answer]
Hi, @Asadbek13. The query builder itself should only mount once, unless you've got some conditional rendering logic higher in the component tree. Similarly, the query builder sub-components will only mount once when they are created through the "+Rule" or "+Group" actions, and then unmount when they are removed by their respective "X" actions.

Did you mean to ask when they are re-rendered? If that's the case, it's a much more complicated answer (see [here](https://blog.isquaredsoftware.com/2020/05/blogged-answers-a-mostly-complete-guide-to-react-rendering-behavior/) and [here](https://www.joshwcomeau.com/react/why-react-re-renders/)) but probably not one you need to worry about. Pretty much any change to the query itself (typing a value, selecting a new field, adding a rule, etc.) will cause the entire query builder component tree to re-render. That doesn't necessarily mean the DOM is updated though, so it's _very_ unlikely to be a performance issue.

Hope that helps.

**Reply by Asadbek13** [reply]
Thank you for your reply! In my case, QueryBuilder is remounting its controllers, and my input in ValueEditor is losing focus every time when its value changes, that's why I was curious about what causes QueryBuilder to remount its controllers. I checked, parent components are not remounted. Any idea about this case?

**Reply by jakeboone02** [reply]
The only time I've seen that before is when a custom `ValueEditor` is defined within another component's render method. See [here](https://react-querybuilder.js.org/docs/tips/common-mistakes#custom-component-as-closure).

**Reply by Asadbek13** [reply]
In my case, ValueEditor is in a separate file. I will try to find the cause but thank you for your time!

**Reply by jakeboone02** [reply]
If you set up a codesandbox, stackblitz, github repo, or anything else that reproduces the problem I'd be happy to take a look.

**Reply by Asadbek13** [reply]
That would be great! Here is the link: https://codesandbox.io/s/querybuilderexample-b8r0gi?file=/src/App.js

**Reply by jakeboone02** [reply]
So it actually was the same problem as I mentioned, but in a roundabout way. When you use arrow function notation _within_ JSX, you're effectively creating a new component on every render even though the component you're rendering is defined in another file.

The difference is `valueEditor: (props) => <MyValueEditor {...props} />` versus `valueEditor: MyValueEditor`. The former creates a new function on every render while the latter only loads the component module once.

You were also doing some funny things with refs and timeouts that, if I'm gleaning your intent correctly, can be avoided by using the [`addRuleToNewGroups` prop](https://react-querybuilder.js.org/docs/api/querybuilder#addruletonewgroups).

[This codesandbox](https://codesandbox.io/s/querybuilderexample-forked-e643fu?file=/src/Editor.js) fixes the re-rendering issue, removes the ref/event/timeout stuff, and adds the `addRuleToNewGroups` prop.

Here is the modified Editor.js for reference:

```jsx
import React from "react";
import QueryBuilder from "react-querybuilder";

import Select from "./Select";
import { generateAttributeMap, prepareQueryBuilderHelpers } from "./utils";
import ValueEditor from "./ValueEditor";
import RemoveButton from "./RemoveButton";

const Editor = ({ query, onChangeFilters, attributes }) => {
  const mapingAttributes = generateAttributeMap(attributes);
  const fields = attributes.map(({ id, display_name }) => ({
    name: id.toString(),
    label: display_name
  }));

  const {
    getValues,
    getValueEditorType,
    getInputType,
    getOperators
  } = prepareQueryBuilderHelpers(mapingAttributes);

  return (
    <QueryBuilder
      query={query}
      fields={fields}
      onQueryChange={onChangeFilters}
      getOperators={getOperators}
      getValueEditorType={getValueEditorType}
      getInputType={getInputType}
      getValues={getValues}
      showCombinatorsBetweenRules
      showNotToggle={false}
      controlClassnames={{ fields: "form-control" }}
      controlElements={{
        valueEditor: ValueEditor,
        combinatorSelector: Select,
        fieldSelector: Select,
        operatorSelector: Select,
        removeRuleAction: RemoveButton,
        removeGroupAction: RemoveButton
      }}
      addRuleToNewGroups
    />
  );
};

export default Editor;
```

**Reply by Asadbek13** [reply]
Thank you very much! It worked.
